support/Python/DirectoryStructure.py

```diff
@@ -0,0 +1,116 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+import logging
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, order=True)
+class Checkpoint:
+    """State of a simulation saved to disk
+
+    An executable can write multiple checkpoints during its execution, and can
+    be restarted from those checkpoints. The 'id' enumerates the checkpoints.
+
+    We currently write checkpoints in a directory structure like this:
+
+    ```
+    RUN_DIR/
+        Checkpoints/
+            Checkpoint000/
+            Checkpoint001/
+            ...
+    ```
+
+    Three-digit IDs give us over a month of hourly checkpoints, or three years
+    of daily checkpoints (usually we checkpoint only once, at the end of a
+    day-long run).
+    """
+    path: Path
+    id: int
+
+    NAME_PATTERN = re.compile(r"Checkpoint(\d+)")
+    NUM_DIGITS = 3
+
+    @classmethod
+    def match(cls, path: Path) -> Optional['Checkpoint']:
+        """Checks if the 'path' is a checkpoint"""
+        match = re.match(cls.NAME_PATTERN, path.resolve().name)
+        if not match:
+            return None
+        return cls(path=path, id=int(match.group(1)))
+
+
+def list_checkpoints(checkpoints_dir: Path) -> List[Checkpoint]:
+    """All checkpoints in the 'checkpoints_dir'"""
+    if not checkpoints_dir.exists():
+        return []
+    matches = map(Checkpoint.match, checkpoints_dir.iterdir())
+    return sorted(match for match in matches if match)
+
+
+@dataclass(frozen=True, order=True)
+class Segment:
+    """Part of a simulation that ran as one executable invocation
+
+    We have to split simulations into segments because supercomputers don't
+    typically support unlimited run times. Therefore, we terminate the job,
+    write the simulation state to disk as a checkpoint, and submit a new job
+    that restarts from the last checkpoint.
+
+    We currently write segments in a directory structure like this:
+
+    ```
+    SEGMENTS_DIR/
+        Segment000/
+            InputFile.yaml
+            Submit.sh
+            Output.h5
+            Checkpoints/
+        Segment001/...
+    ```
+
+    With SpEC we never needed more than 26*26=676 segments (AA-ZZ), so three
+    digits for segments is enough.
```

Comment:
Is this true?  I vaguely think I've seen three-character SpEC segment IDs.

Reply:
I haven't seen those, but that doesn't say much.

Three digits give us over 3 years in daily segments. That seems enough for the vast majority of simulations. As mentioned above the number of digits is not set in stone, so if we want to do a 5-year simulation or do hourly segments for some reason or something like that we can always run with more digits.